api/jobs.go

```diff
@@ -852,29 +856,48 @@ func (p *PeriodicConfig) Canonicalize() {
 // passed time.
 func (p *PeriodicConfig) Next(fromTime time.Time) (time.Time, error) {
 	if p != nil && *p.SpecType == PeriodicSpecCron {
-		e, err := cronexpr.Parse(*p.Spec)
-		if err != nil {
-			return time.Time{}, fmt.Errorf("failed parsing cron expression %q: %v", *p.Spec, err)
+		if p.Spec != nil && *p.Spec != "" {
+			return cronParseNext(fromTime, *p.Spec)
 		}
-		return cronParseNext(e, fromTime, *p.Spec)
-	}
 
+		if p.Specs != nil && len(p.Specs) != 0 {
+			times := make([]time.Time, len(p.Specs))
+			var nextTime time.Time
+			var err error
+			for i, spec := range p.Specs {
+				times[i], err = cronParseNext(fromTime, spec)
+				if err != nil {
+					return time.Time{}, fmt.Errorf("failed parsing cron expression %s: %v", spec, err)
+				}
+			}
+			nextTime = times[0]
+			for _, next := range times {
+				if next.Before(nextTime) {
+					nextTime = next
+				}
+			}
+			return nextTime, nil
+		}
```

Comment:
```suggestion
		var nextTime time.Time
		for _, spec := range p.Specs {
			t, err := cronParseNext(fromTime, spec)
			if err != nil {
				return time.Time{}, fmt.Errorf("failed parsing cron expression %s: %v", spec, err)
			}
			if nextTime.IsZero() || t.Before(nextTime) {
				nextTime = t
			}
		}
		return nextTime, nil
```

Sorry, I should've noticed this before, but I think we can avoid iterating over the times twice since we only care about keeping the earliest next time.

And since we don't have a pointer to a slice anymore we remove the `if` check and the `range` will be safe, even if `Specs` is `nil`.

Reply:
```golang
if nextTime.IsZero() || t.Before(nextTime) {
			nextTime = t
		}
```
Very good advice! Sorry for I skip it.